lib/zip_source_zip_new.c

```diff
@@ -100,29 +132,50 @@ zip_source_t *_zip_source_zip_new(zip_t *srcza, zip_uint64_t srcidx, zip_flags_t
     }
     _zip_file_attributes_from_dirent(&attributes, de);
 
-    if (st.comp_size == 0) {
+    if (st.size == 0) {
         return zip_source_buffer_with_attributes_create(NULL, 0, 0, &attributes, error);
     }
 
+    src = srcza->src;
+    source_archive = srcza;
+    source_index = srcidx;
+    if (reopening) {
+        src = srcza->entry[srcidx].source;
+        source_archive = NULL;
+        source_index = 0;
+    }
     if (partial_data && !needs_decrypt && !needs_decompress) {
         struct zip_stat st2;
 
         st2.size = len;
         st2.comp_size = len;
         st2.comp_method = ZIP_CM_STORE;
-        st2.mtime = st.mtime;
-        st2.valid = ZIP_STAT_SIZE | ZIP_STAT_COMP_SIZE | ZIP_STAT_COMP_METHOD | ZIP_STAT_MTIME;
+        st2.valid = ZIP_STAT_SIZE | ZIP_STAT_COMP_SIZE | ZIP_STAT_COMP_METHOD;
+        if (st.valid & ZIP_STAT_MTIME) {
+            st2.mtime = st.mtime;
+            st2.valid |= ZIP_STAT_MTIME;
+        }
 
-        if ((src = _zip_source_window_new(srcza->src, start, (zip_int64_t)len, &st2, &attributes, srcza, srcidx, error)) == NULL) {
+        if ((src = _zip_source_window_new(src, start, (zip_int64_t)len, &st2, &attributes, source_archive, source_index, error)) == NULL) {
             return NULL;
         }
     }
     else {
```

Comment:
This else branch only applies if not reopening: It restricts src to the file data, while srcza->src is the complete archive. When reopening, src is already just the file data.

Reply:
I'll make if `else if (!changed_data) {` then and add a comment here. Essentially this branch will be only executed for `srcza->src`, so should I revert the changes I made for this branch? (Can we assume that `st` is going to have a valid `comp_size` member set?)

Reply:
Yes, please revert your changes here. You can assume that `comp_size` is valid here, since the stat information is taken from the archive.